### PR TITLE
radicle-surf: Namespace using RefString

### PR DESCRIPTION
**Pull request by FintanH**
The main change is described in https://github.com/radicle-dev/radicle-git/commit/e4853c10bcc5606ea419f07b6f043a4b280b88b3 but required a few changes in `git-ref-format`.

### CODE REVIEW

File: radicle-surf/src/git/glob.rs [hunk 1]
```diff
@@ -28,22 +33,20 @@ pub enum Error {
...
-    globs: Vec<PatternString>,
+    globs: Vec<QualifiedPattern<'static>>,
     glob_type: PhantomData<T>, // To support different methods for different T.
 }
 
 impl<T> Glob<T> {
-    /// Returns the globs.
-    pub fn globs(&self) -> Vec<&str> {
-        self.globs.iter().map(|g| g.as_str()).collect()
+    pub fn globs(self) -> impl Iterator<Item = QualifiedPattern<'static>> {
```
Comment (keepsimple1) [inline radicle-surf/src/git/glob.rs:41]: Do we have to consume `self`?

File: radicle-surf/src/git/glob.rs [hunk 2]
```diff
@@ -52,16 +55,43 @@ impl<Namespace> Glob<Namespace> {
...
-        self.globs.push(pattern);
+        self.globs.push(Self::qualify(glob)?);
         Ok(self)
     }
+
+    fn qualify(glob: &str) -> Result<QualifiedPattern<'static>, Error> {
+        Ok(refname!("refs/namespaces")
+            .to_pattern(PatternString::try_from(glob)?)
+            .qualified()
+            .expect("BUG: pattern is qualified")
```
Comment (keepsimple1) [inline radicle-surf/src/git/glob.rs:66]: I think it's better to return an error here instead of panic.

File: radicle-surf/src/git/glob.rs [hunk 3]
```diff
@@ -52,16 +55,43 @@ impl<Namespace> Glob<Namespace> {
...
+impl FromIterator<PatternString> for Glob<Namespace> {
+    fn from_iter<T: IntoIterator<Item = PatternString>>(iter: T) -> Self {
+        let globs = iter
+            .into_iter()
+            .map(|pat| {
+                refname!("refs/namespaces")
+                    .to_pattern(pat)
+                    .qualified()
+                    .expect("BUG: pattern is qualified")
+                    .into_owned()
```
Comment (keepsimple1) [inline radicle-surf/src/git/glob.rs:80]: nit: this is the same logic as in `qualify` above, is it possible to extract it into a function and call the function in both places?

File: radicle-surf/src/git/glob.rs [hunk 4]
```diff
@@ -71,17 +101,43 @@ impl<Tag> Glob<Tag> {
...
+        self.globs.push(Self::qualify(glob)?);
         Ok(self)
     }
+
+    fn qualify(glob: &str) -> Result<QualifiedPattern<'static>, Error> {
+        Ok(refname!("refs/tags")
+            .to_pattern(PatternString::try_from(glob)?)
+            .qualified()
+            .expect("BUG: pattern is qualified")
+            .into_owned())
```
Comment (keepsimple1) [inline radicle-surf/src/git/glob.rs:113]: same here. Is it possible to call a common function for the above 5 lines?

File: radicle-surf/src/git/glob.rs [hunk 5]
```diff
@@ -71,17 +101,43 @@ impl<Tag> Glob<Tag> {
...
+impl FromIterator<PatternString> for Glob<Tag> {
+    fn from_iter<T: IntoIterator<Item = PatternString>>(iter: T) -> Self {
+        let globs = iter
+            .into_iter()
+            .map(|pat| {
+                refname!("refs/tags")
+                    .to_pattern(pat)
+                    .qualified()
+                    .expect("BUG: pattern is qualified")
+                    .into_owned()
```
Comment (keepsimple1) [inline radicle-surf/src/git/glob.rs:126]: and here to call the same function ?

File: radicle-surf/src/git/glob.rs [hunk 6]
```diff
@@ -100,15 +155,29 @@ impl<Branch> Glob<Branch> {
...
+        self.globs.push(Self::qualify_remotes(glob)?);
         Ok(self)
     }
+
+    fn qualify_heads(glob: &str) -> Result<QualifiedPattern<'static>, Error> {
+        Ok(refname!("refs/heads")
+            .to_pattern(PatternString::try_from(glob)?)
+            .qualified()
+            .expect("BUG: pattern is qualified")
+            .into_owned())
```
Comment (keepsimple1) [inline radicle-surf/src/git/glob.rs:173]: same here for calling a common function.

File: radicle-surf/src/git/glob.rs [hunk 7]
```diff
@@ -100,15 +155,29 @@ impl<Branch> Glob<Branch> {
...
+    }
+
+    fn qualify_remotes(glob: &str) -> Result<QualifiedPattern<'static>, Error> {
+        Ok(refname!("refs/remotes")
+            .to_pattern(PatternString::try_from(glob)?)
+            .qualified()
+            .expect("BUG: pattern is qualified")
+            .into_owned())
```
Comment (keepsimple1) [inline radicle-surf/src/git/glob.rs:181]: same here for calling a common function.

**Review by keepsimple1 — COMMENTED** [review]
Overall looks good to me. Thanks. A few comments inline.

File: radicle-surf/src/git/glob.rs (hunk 2 above)
Comment (FintanH) [reply radicle-surf/src/git/glob.rs:66]: But it can't panic because it's qualified by definition here.

**Pull request merged**